### ISSUE TITLE
404 when speaking to private registry

### ISSUE THREAD

**Issue by btrepp** [updatecli/updatecli#600]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Using
```
sources:
  registry:
    kind: dockerDigest
    spec:
      image: "registry.concernedturtle.com/btrepp/blog"
      architecture: arm64
      username: beau
      password: gibberish
      
 ```

With
```
updatecli diff --debug
``` 

Gives
 ```
SOURCES
=======

registry
--------
DEBUG: Searching digest for the image "registry.concernedturtle.com/btrepp/blog" with the tag ""
DEBUG: Retrieving a bearer token from https://registry.concernedturtle.com/token?scope=repository:btrepp/blog:pull&service=registry.concernedturtle.com with the username beau and the associated password.
DEBUG: Emitting a request to https://registry.concernedturtle.com/token?scope=repository:btrepp/blog:pull&service=registry.concernedturtle.com
ERROR: Got an HTTP response code 404 while logging in to the registry registry.concernedturtle.com.
ERROR: ✗ Got an HTTP response code 404 while logging in to the registry registry.concernedturtle.com.
Pipeline "UPDATECLI.YAML" failed
Skipping due to:
        "sources stage:\t\"Got an HTTP response code 404 while logging in to the registry registry.concernedturtle.com.\""
sources stage:  "Got an HTTP response code 404 while logging in to the registry registry.concernedturtle.com."

```

and logs on the registry

```
10.42.1.49 - - [21/Mar/2022:10:27:39 +0000] "GET /token?scope=repository:btrepp/blog:pull&service=registry.concernedturtle.com HTTP/1.1" 404 19 "" "Go-http-client/2.0"
```

### Expected Behavior

Update cli connects to the registry with correct password.
Or in the case of that incorrect password above gives 401 instead of 404.

Note it gives 404 with both correct and incorrect passwords.

### Steps To Reproduce

See above expected output

### Environment

```markdown
- OS:

uname -a
Linux beau-pc 5.16.11-1-MANJARO-ARM #1 SMP PREEMPT Wed Feb 23 12:30:41 UTC 2022 aarch64 GNU/Linux
```
- updatecli:
```
updatecli version

VERSION

Application:    0.22.2
Golang     :    1.17.8 linux/amd64
Build Time :    2022-03-20T18:54:44Z
```

### Anything else?

It would be great if the credentials for registrys could be pulled from 
```
~/.docker/config.json
```

This would help in troubleshooting a fair bit, as if you can login with docker cli, that means your auth would be setup. I'm not 100% clear if I should use username/password, or token with the base64 encoding in the config.json file.

I am also not sure if this may be a feature request?. As private registries may not be supported.

**Comment by dduportal** [comment]
Hello @btrepp thanks for reporting. This is definitively a feature request (and not a bug) as we never tried a local private registry: it needs a bit of discussion to ensure that we shape the specification. I'm changing the labelling of the issue.

**Comment by dduportal** [comment]
Lets discuss :) 

A few elements about how the docker registries usually handle the bearer token authentication:

- The usage of username + password is mandatory. You might see the concept of a "token" in SaaS hosted registries such as the DockerHub: in these case you specify the token as "password": I'll refer to this token as an "API token".
  - We might need to ensure that the documentation of updatecli helps the end users.
  
- As per the Docker Registry v2 API, the "bearer" token is supported: https://docs.docker.com/registry/spec/auth/token/
  - This "bearer token" is generated from a first HTTP request to the authentication endpoint with the username + password (or username + API token), which responds a "bearer token" to be used for later authentications. Please note that the bearer token has an expiration date (while the API token does not, like a password).
  - Then, for any subsequent request to the registry itself, the bearer token is provided (instead of username + password) to authenticate these request. The registry would use the authentication service as a "3rd party" to validate the authenticity of the bearer token.

In you case @btrepp , the "HTTP/404"  error is because there isn't any authentication service that would provide a bearer token: authentication is managed directly with (username + password) by your registry, out of the box:

- You correctly configured updatecli! And it should stay the same configuration for at least username password
- There is no such thing as an API token in your case 
- I'm not sure about the role of tokens found in `~/.docker/config.json`. Can you check (without sharing its content of course, or anonymized) if the tokens aren't related to another registry such as dockerhub?


To implement support for a private registry, we need to think about a way to sepcify, in the YAML manifest, that there should not be any "authentication endpoint" (named "token endpoint" in the code) + that username+password should be passed "as -it" instead of expecting a bearer token.

**Comment by btrepp** [comment]
No worries. Figured it was most likely unsupported, but couldn't make a call if it was my config or a missing feature. So thanks for confirming, and this issue can be the documentation that it's not implemented yet for anyone else :).

One the docker config. I think this might be a seperate feature too, but it does help explain some of the situation, so will mention in this thread.

The file looks like
```
{
        "auths": {
                "https://index.docker.io/v1/": {
                        "auth": "REDACTED"
                },
                "registry.concernedturtle.com": {
                        "auth": "REDACTED"
                }
        }
}
```

I generate one for m k8s cluster here https://git.sr.ht/~btrepp/infrastructure/tree/main/item/bootstrap/playbooks/secrets/dockerconfig.json.j2
```
{
    "auths": {
        "registry.concernedturtle.com": {
            "auth": "{{ ('k8s1:'+registry_k8s1_pass) | b64encode }}"
        }
    }
}
```
So it is the `username : password` then base64 encoded. 

This all works nicely in k8s and docker cli. 

With the auth spec, I think the most confusing part, is that I am running the https://hub.docker.com/_/registry which you would think implements that specification you have linked.

Thinking deeper, perhaps it's that my basic configuration needs more switches turned on https://git.sr.ht/~btrepp/infrastructure/tree/main/item/apps/registry/deployment.yaml. I've noticed that says htpasswd as auth, perhaps I need to enable something else too.
My thinking here, is that this should be the same as the public registry, it does implement the spec. 

The feature request for updatecli, might be 'support registrys that use htpasswd auth'. I think in that case, maybe an enum field, that says 'authentication_method': [TOKEN,HTPASSWD]. Might work?.
I suspect the docker tooling 'figures this out', but allowing users to specify would be a good MVP.

I will also do some research to see if I can 'turn on' the registry auth to work similar as expected. If that works, maybe no changes are needed, just update the documentation to say we need these bits turned on :).

**Comment by btrepp** [comment]
Basic/htpassswd seems to be documented as the a good default for  private installs. The JWT token path actually is more undefined :).
https://github.com/distribution/distribution/blob/main/docs/deploying.md

I think flow wise, perhaps when getting the 404, try using Basic auth for the rest of the requests (no need to document more). I think this is probably what the docker/k8s tooling does, but maybe I can find the source code for them

**Comment by olblak** [comment]
As far as I understood this is issue solved in by https://github.com/updatecli/updatecli/pull/900 which was released in 0.36.0
